hydra-main/src/main/java/com/addthis/hydra/job/alias/AliasManagerImpl.java

```diff
@@ -51,14 +99,158 @@ public AliasManagerImpl(SpawnDataStore spawnDataStore) {
+        } catch (Exception e) {
+            log.error("",e);
+        }
+    }
+
+    /**
+     * Load the jobIds for a particular alias from the SpawnDataStore
+     *
+     * @param alias The alias key to check
+     * @return String The data that was updated (so the cache can be updated)
+     */
+    @Nullable private String updateAlias(String alias, @Nullable String data) {
+        if (Strings.isNullOrEmpty(alias)) {
+            return data;
+        }
+        if (Strings.isNullOrEmpty(data)) {
+            deleteAlias(alias);
+            return data;
+        }
+        List<String> jobs = decodeAliases(data);
+        if (jobs.isEmpty()) {
+            log.warn("no jobs for alias {}, ignoring {}", alias, alias);
+            return data;
+        }
+        mapLock.lock();
+        try {
+            alias2jobs.put(alias, jobs);
+            job2alias.put(jobs.get(0), alias);
+        } finally {
+            mapLock.unlock();
+        }
+        return data;
+    }
+
+    @VisibleForTesting
+    protected List<String> decodeAliases(@Nonnull Object data) {
+        try {
+            return mapper.readValue(data.toString(), new TypeReference<List<String>>() {});
+        } catch (IOException e) {
+            log.warn("Failed to decode data", e);
+            return new ArrayList<>(0);
+        }
+    }
+
+    /**
+     * Delete the data for a given alias
+     *
+     * @param alias The alias to check
+     */
     public void deleteAlias(String alias) {
-        aliasBiMap.deleteAlias(alias);
+        mapLock.lock();
+        try {
+            List<String> jobs = alias2jobs.get(alias);
+            alias2jobs.remove(alias);
+            if ((jobs != null) && !jobs.isEmpty()) {
+                for (String job : jobs) {
+                    String aliasVal = job2alias.get(job);
+                    if (Objects.equals(aliasVal, alias)) {
+                        job2alias.remove(job);
+                    }
+                }
+            }
+        } finally {
+            mapLock.unlock();
+        }
+        spawnDataStore.deleteChild(ALIAS_PATH, alias);
+        ac.deleteAlias(alias);
+    }
+
+    /**
+     * Get an alias for a particular jobId
+     *
+     * @param jobid The jobId to check
+     * @return One of the aliases for that job
+     */
+    public String getLikelyAlias(String jobid) {
+        mapLock.lock();
+        try {
+            String tmpAlias = job2alias.get(jobid);
+            if (tmpAlias != null) {
+                // Check to see if the alias has been deleted
+                checkAlias(jobid, tmpAlias);
+            }
+            return job2alias.get(jobid);
+        } finally {
+            mapLock.unlock();
+        }
     }
 
+    /**
+     * Test a job/alias pair to see if an alias has disappeared
+     *
+     * @param job   The job to test
+     * @param alias The alias to check
+     */
+    private void checkAlias(String job, String alias) {
```

Comment:
This method also has unclear side effects since it removes job from `job2alias` when the method name is called `checkAlias`